rpc: deprecate top-level fee fields in getmempool RPCs

josibake opened the pull request:
per #22682 , top level fee fields for mempool entries have been deprecated since 0.17 but are still returned. this PR properly deprecates them so that they are no longer returned unless `-deprecatedrpc=fees` is passed.

the first commit takes care of deprecation and also updates `test/functional/mempool_packages.py` to only use the `fees` object. the second commit adds a new functional test for `-deprecatedrpc=fees`

closes #22682

## questions for the reviewer

* `-deprecatedrpc=fees` made the most sense to me, but happy to change if there is a name that makes more sense
* #22682 seems to indicate that after some period of time, the fields will be removed all together. if we have a rough idea of when this will be, i can add a `TODO: fully remove in vXX` comment to `entryToJSON`

## testing
to get started on testing, compile, run the tests, and start your node with the deprecated rpcs flag:

```bash
./src/bitcoind -daemon -deprecatedrpc=fees
```
you should see entries with the deprecated fields like so:
```json
{
  "<txid>": {
    "fees": {
      "base": 0.00000671,
      "modified": 0.00000671,
      "ancestor": 0.00000671,
      "descendant": 0.00000671
    },
    "fee": 0.00000671,
    "modifiedfee": 0.00000671,
    "descendantfees": 671,
    "ancestorfees": 671,
    "vsize": 144,
    "weight": 573,
   ...
  },
```
you can also check `getmempoolentry` using any of the txid's from the output above. 

next start the node without the deprecated flag, repeat the commands from above and verify that the deprecated fields are no longer present at the top level, but present in the "fees" object

josibake committed:
rpc: deprecate fee fields from mempool entries

Unless `-deprecatedrpc=fees` is passed, top level
fee fields are no longer returned for mempool entries.

Add instructions to field help on how to access
deprecated fields, update help text for readability,
and include units. This is important to help
avoid any confusion as users move from deprecated
fields to the fee fields object (credit: jonatack).

This affects `getmempoolentry`, `getrawmempool`,
`getmempoolancestors`, and `getmempooldescendants`

Modify `test/functional/mempool_packages.py` and
`test/functional/rpc_fundrawtransaction.py` tests
to no longer use deprecated fields.

Co-authored-by: jonatack <jon@atack.com>

josibake committed:
test: add functional test for deprecatedrpc=fees

Test for old fields when `-deprecatedrpc=fees`
flag is passed and verify values in the deprecated
fields match values in the fees sub-object.

josibake commented:
force pushed from  084d443 to 251d636 to fix linting and ci errors

glozow commented on src/rpc/blockchain.cpp:
```diff
@@ -481,7 +481,7 @@ static std::vector<RPCResult> MempoolEntryDescription() { return {
 };}
 
-static void entryToJSON(const CTxMemPool& pool, UniValue& info, const CTxMemPoolEntry& e) EXCLUSIVE_LOCKS_REQUIRED(pool.cs)
+static void entryToJSON(const CTxMemPool& pool, UniValue& info, const CTxMemPoolEntry& e, bool include_fee_fields) EXCLUSIVE_LOCKS_REQUIRED(pool.cs)
```
I agree with this approach of adding a bool similar to address/reqSigs, and think you should add a TODO to remove this in v24

Perhaps naming it `include_old_fee_fields` or something might be better because this could be slightly confusing

josibake replied:
added the comments and renamed to `include_deprecated_fee_fields` to better associate it with `isDeprecatedRPCEnabled("fees")`

glozow commented on test/functional/rpc_mempool_entry_fee_fields_deprecation.py:
```diff
@@ -0,0 +1,62 @@
+# Copyright (c) 2021 The Bitcoin Core developers
+# Distributed under the MIT software license, see the accompanying
+# file COPYING or http://www.opensource.org/licenses/mit-license.php.
+"""Test deprecation of fee fields from top level mempool entry object"""
```
Any reason you didn't use the existing test file, rpc_deprecated.py? I don't personally have an opinion on whether it would be better, but it looks like a good place to put these tests too.

josibake replied:
i saw `rpc_deprecated.py` and wasn't sure why it was there? right now it doesn't test anything. i then saw `rpc_addresses_deprecation.py` was its own file, so i followed that one as an example.

i guess the advantage of putting them there is it might help things run faster than running individual test files for each deprecated rpc. i can put it there and then move `rpc_addresses_deprecation.py` to `rpc_deprecated.py` as a follow-up pr

glozow replied:
I assumed that rpc_deprecated.py there to test RPC deprecation stuff and address/reqSigs tests are in a separate file because they need the wallet to be built. So they can't be moved to rpc_deprecated.py either.

josibake replied:
actually, ive talked myself out of this. if we have individual files, they can run in parallel and if we put all the deprecated tests under the same class it could be really annoying later on as you are adding a new test. you'd have to wait for N deprecation tests to run each time as you're iterating on your test, which starts to impact productivity if it takes 30 ~ 45 seconds to run through all the tests each time.

what are your thoughts, @glozow ?

josibake replied:
> I assumed that rpc_deprecated.py is there to test RPC deprecation stuff and address/reqSigs tests are in a separate file because they need the wallet to be built

makes sense. i read through the comments on `rpc_deprecated.py` and it seems focused on testing that a whole rpc command is deprecated. i think it works well for that and we can consolidate rpc command deprecation's there.

this is test slightly different in that an rpc is not being deprecated (meaning we cant just check for an rpc not found error), but fields on an rpc response are being deprecated, which makes the setup for the test a bit more complicated. for this reason, id suggest leaving this in a separate file

glozow replied:
I don't think it matters that much :shrug: either way makes sense to me. I'd imagine that the extra overhead of spinning up test frameworks for separate test files overrides the gains from running them in parallel.

glozow commented on test/functional/rpc_mempool_entry_fee_fields_deprecation.py:
```diff
@@ -0,0 +1,62 @@
+        txid = self.wallet.send_self_transfer(from_node=self.nodes[0])['txid']
+        self.sync_mempools(nodes=self.nodes)
+
+        # note: we are only testing getmempoolentry even though
+        # getrawmempool(verbose=True) also returns mempool entries
+        # this is because getmempool entry and getrawmempool(verbose=True)
+        # both call the same underlying entryToJSON function so testing
+        # getmempoolentry is sufficient
```
You could also add the assertions to a helper function, and call that helper function with results you get from each RPC

glozow commented on test/functional/rpc_mempool_entry_fee_fields_deprecation.py:
```diff
@@ -0,0 +1,62 @@
+# file COPYING or http://www.opensource.org/licenses/mit-license.php.
+"""Test deprecation of fee fields from top level mempool entry object"""
+
+from test_framework.wallet import MiniWallet
+from test_framework.test_framework import BitcoinTestFramework
+from test_framework.util import assert_equal
+from test_framework.blocktools import COIN
```
nit: sort alphabetically :grin:

josibake replied:
:doh:

glozow commented on test/functional/rpc_fundrawtransaction.py:
```diff
@@ -379,7 +379,7 @@ def test_fee_p2pkh(self):
         # Create same transaction over sendtoaddress.
         txId = self.nodes[0].sendtoaddress(self.nodes[1].getnewaddress(), 1.1)
-        signedFee = self.nodes[0].getrawmempool(True)[txId]['fee']
+        signedFee = self.nodes[0].getrawmempool(True)[txId]['fees']['base']
```
To make the commits atomic, rpc_fundrawtransaction.py needs to be updated in 5adcba2562 as well

glozow replied:
ah hm, I must have fetched before you pushed.

> just to make sure i understand, the changes to rpc_fundrawtransaction.py and mempool_packages.py should be in the same commit as the changes to blockchain.cpp. this ensures the tests still pass for this commit, correct?

correct, yes :)

glozow reviewed (commented):
Concept and approach ACK

theStack commented:
Concept ACK

Zero-1729 commented:
Concept ACK

josibake commented:
thanks for the review, @glozow! I added your suggestions and left a comment on why I think it makes sense to keep `rpc_mempool_entry_fields_deprecation.py` separate from `rpc_deprecated.py`; ultimately I'm happy to go with whichever option you recommend

mjdietzx commented on src/rpc/blockchain.cpp:
```diff
@@ -491,19 +492,20 @@ static void entryToJSON(const CTxMemPool& pool, UniValue& info, const CTxMemPool
     fees.pushKV("descendant", ValueFromAmount(e.GetModFeesWithDescendants()));
     info.pushKV("fees", fees);
-
+    if (include_deprecated_fee_fields) {
```
Is there a reason you don't do:
```c++
if (IsDeprecatedRPCEnabled("fees")) {
```
here? And not have to change the function header?

josibake replied:
i was following `ScriptPubKeyToUniv` as an example: https://github.com/bitcoin/bitcoin/blob/42b00a37580a11fbb81a3748d9e834ae8913cdba/src/rpc/blockchain.cpp#L1936-L1939

personally, i agree with this approach more than using the `IsDeprecatedRPCEnabled` function directly as it it's more clear, imo, what's happening and allows for more flexibility. 

for example, lets say we wanted to only allow `getmempoolentry` to return the fields when `-deprecated=fees` is passed but wanted `getrawmempool` to never return the fields. with this setup, we can have one of the functions use the original function signature with `IsDeprecated` passed as the default, whereas we could call `entryToJSON` with `false` passed as the default whenever it's used in `getrawmempool`.

tldr; i think functional overloading is super cool and this seemed a good use case for it :smile:

MarcoFalke replied:
As this `entryToJSON` isn't called anywhere, except for `entryToJSON`, I think it is fine to use the suggestion by @mjdietzx . No strong opinion, though.

jonatack replied:
I agree as well, but no big deal so didn't mention it.

josibake replied:
seems like its the more preferred approach, happy to switch

mjdietzx reviewed (commented):
Code review ACK

You'll probably need to add release notes, something like 90ae3d8ca68334ec712d67b21a8d4721c6d79788

josibake committed:
doc: add release note for fee field deprecation

josibake commented:
> Code review ACK
> 
> You'll probably need to add release notes, something like [90ae3d8](https://github.com/bitcoin/bitcoin/commit/90ae3d8ca68334ec712d67b21a8d4721c6d79788)

thanks for the review, @mjdietzx ! ive updated the release notes in https://github.com/bitcoin/bitcoin/pull/22689/commits/4df6ffd4dcdfd2b26457c420cdbb4c382bc7a488

fanquake commented:
> thanks for the review, @mjdietzx ! ive updated the release notes in 4df6ffd

Please write more descriptive commit messages than `update release notes`. i.e This could be something like `doc: add release note for fee field deprecation in RPC`.

josibake commented:
> Please write more descriptive commit messages than `update release notes`. i.e This could be something like `doc: add release note for fee field deprecation in RPC`.

great callout, @fanquake . ive update all three commit messages with relevant tags (rpc, test, doc) and wrote more descriptive commit messages

jonatack commented on doc/release-notes.md:
```diff
@@ -63,6 +63,8 @@ P2P and network changes
 Updated RPCs
 ------------
+- `getmempoolentry` and `getrawmempool true` no longer return top level fee fields `fee`, `modifiedfee`, `ancestorfees`, `descendantfees`. These fields were deprecated in `0.17` and moved to the `fees` sub-object in the response.
+The `-deprecated=fees` flag must be passed for these fields to be included in the top level response. Note: this flag will only be available until `v24`, after which the fields will be fully removed. It is recommended you migrate to using the `fees` sub-object in the response
```
suggestions

- add the PR number
- add  `getmempoolancestors` and `getmempooldescendants` that are also changed
- s/true/verbose=true/ for clarity
- wrap the lines
- s/sub-object/JSON object/
- s/top level/top-level/
- a few others

```diff
-- `getmempoolentry` and `getrawmempool true` no longer return top level fee fields `fee`, `modifiedfee`, `ancestorfees`, `descendantfees`. These fields were deprecated in `0.17` and moved to the `fees` sub-object in the response.
-The `-deprecated=fees` flag must be passed for these fields to be included in the top level response. Note: this flag will only be available until `v24`, after which the fields will be fully removed. It is recommended you migrate to using the `fees` sub-object in the response
+- RPCs `getmempoolentry`, `getrawmempool verbose=true`, `getmempoolancestors
+  verbose=true` and `getmempooldescendants verbose=true` no longer return the
+  top-level fee fields `fee`, `modifiedfee`, `ancestorfees` and
+  `descendantfees`. These fields were deprecated in v0.17 and moved to the
+  `fees` JSON object in the response. The `-deprecated=fees` configuration
+  option must be passed for these fields to be included in the top-level
+  response. This option will only be available until v24.0, after which the
+  fields will be fully removed. Use the `fees` JSON object for these RPCS
+  instead. (#22689)
```

jonatack reviewed (commented):
Code review ACK modulo the comment below and updating the first commit message, as getmempoolancestors and getmempooldescendants are also changed unless I'm misreading the code

Edit: perhaps rename the PR to "rpc: deprecate top-level fee fields in getmempool RPCs"

josibake commented:
> getmempoolancestors and getmempooldescendants are also changed unless I'm misreading the code

good catch! i verified by grepping for `entryToJSON` and it is used in those RPC's as well. updated the commit message and release docs. thanks for the suggestions on the release doc; it reads much better

josibake commented:
force pushed https://github.com/bitcoin/bitcoin/pull/22689/commits/e5b22ecc858f03be29a8f5aa1802b679a782a593 to fix a failure in the linter

josibake commented:
force pushed https://github.com/bitcoin/bitcoin/pull/22689/commits/07801c89d6782d404370a52566a55375f36846d7 based on https://github.com/bitcoin/bitcoin/pull/22689#discussion_r689259509

MarcoFalke commented on doc/release-notes.md:
```diff
@@ -63,6 +63,15 @@ P2P and network changes
+  `descendantfees`. These fields were deprecated in v0.17 and moved to the
+  `fees` JSON object in the response. The `-deprecated=fees` configuration
+  option must be passed for these fields to be included in the top-level
+  response. This option will only be available until v24.0, after which the
```
nit: It won't be available in v24.0

```suggestion
  response. This option will only be available in this major version, after which the
```

MarcoFalke commented on src/rpc/blockchain.cpp:
```diff
@@ -481,6 +481,7 @@ static std::vector<RPCResult> MempoolEntryDescription() { return {
     RPCResult{RPCResult::Type::BOOL, "unbroadcast", "Whether this transaction is currently unbroadcast (initial broadcast not yet acknowledged by any peers)"},
 };}
 
+// TODO: top level fee fields are deprecated. this method should be refactored in v24 to completely remove them
```
nit: Could move the comment to the `if IsDeprecatedRPCEnabled`?

MarcoFalke commented on test/functional/mempool_packages.py:
```diff
@@ -194,11 +188,9 @@ def run_test(self):
-                assert_equal(mempool[x]['modifiedfee'], mempool[x]['fee']+satoshi_round(0.00002))
-                assert_equal(mempool[x]['fees']['modified'], mempool[x]['fee']+satoshi_round(0.00002))
-            assert_equal(mempool[x]['descendantfees'], descendant_fees * COIN + 2000)
+                assert_equal(mempool[x]['fees']['modified'], mempool[x]['fees']['base']+satoshi_round(0.00002))
```
nit: pep-8

```suggestion
                assert_equal(mempool[x]['fees']['modified'], mempool[x]['fees']['base'] + satoshi_round(0.00002))
```

maflcko reviewed (approved):
cr ACK b409e33c682a37ecd965b6d9dc2cae35ee672b59

josibake commented:
thanks for the review, @MarcoFalke , ive updated based on your suggestions

MarcoFalke commented:
re-ACK a0a0c7c1c8e41157bb2798ba0d564e78e5217ff9

mjdietzx reviewed (commented):
reACK a0a0c7c1c8e41157bb2798ba0d564e78e5217ff9 well done 👍

jonatack commented on src/rpc/blockchain.cpp:
```diff
@@ -491,19 +491,21 @@ static void entryToJSON(const CTxMemPool& pool, UniValue& info, const CTxMemPool
+        info.pushKV("modifiedfee", ValueFromAmount(e.GetModifiedFee()));
+        info.pushKV("descendantfees", e.GetModFeesWithDescendants());
+        info.pushKV("ancestorfees", e.GetModFeesWithAncestors());
+    }
```
Might be more user-friendly to not alter the field order here so that it is the same as the help. E.g. insert the ISDeprecated... conditional (or cache the result of the call to a local bool variable) before the fields where they already are.

MarcoFalke replied:
The code will go away soon, so I don't think this gives much benefit.

josibake replied:
i noticed that the current order doesn't match the help (fees object is listed after wtxid but is first in the returned object), so didn't worry about re-arranging fields, but i agree that it can be confusing for the user. 

i think grouping the fields makes it easier to deprecate them later, so maybe a better solution would be to re-arrange the help to match the actual ordering? or just leave it as is per @MarcoFalke 's comment, i dont have a strong opinion either way

jonatack replied:
I would say, please don't re-arrange the help to make up for something the code can easily do.

Yes, I agree it's not of earth-shattering importance, but I also think it's a more important aspect than in which file to put the deprecation test or which way to check for deprecation. It affects users.

jonatack replied:
> The code will go away soon, so I don't think this gives much benefit.

Same for the deprecation test and which way to check for deprecation, but the pull was updated for these. I guess it's a question of degree of focus on users. Either way, not a big deal.

josibake replied:
> I guess it's a question of degree of focus on users

good point, @jonatack , i'll update based on your suggestion

jonatack replied:
Happy to re-review ASAP if you do (it's not a blocker).

jonatack replied:
> i noticed that the current order doesn't match the help (fees object is listed after wtxid but is first in the returned object)

Oh, I see now what you meant. Hm, the fees output was done in 7de1de7da49 in 2018 and the fees help in fa6b061fc11 in 2020. It looks like the later commit auto-moved JSON objects after top-level fields? So your change ~seems correct, unless @glozow disagrees whether better to change the help to be like the output.

jonatack replied:
> i noticed that the current order doesn't match the help (fees object is listed after wtxid but is first in the returned object)

Oh, I see now what you meant. Hm, the fees output/help mismatch was done in 7de1de7da49. ISTM we usually place JSON objects after top-level fields? So your change ~seems correct, unless other reviewers think it's better to change the help to be like the output.

josibake replied:
> JSON objects after top-level fields

that's the convention i'm used to (and much prefer)

josibake committed:
rpc: move fees object to match help

josibake commented:
pushed changes per @jonatack 's suggestion to preserve field order to match rpc help. i also added https://github.com/bitcoin/bitcoin/pull/22689/commits/c1a8a35e89d5b3c889d64562c2e5f042c16a4ae2 to make the fees object also match the help. given the discussion in https://github.com/bitcoin/bitcoin/pull/22689#discussion_r692053229 , this seemed like a good change

i did it in a separate commit because it is _slightly_ unrelated to deprecating the fields and i can easily drop the commit if people feel its too out of scope.

glozow commented on doc/release-notes.md:
```diff
@@ -63,6 +63,15 @@ P2P and network changes
 
 Updated RPCs
 ------------
+- RPCs `getmempoolentry`, `getrawmempool verbose=true`, `getmempoolancestors
+  verbose=true` and `getmempooldescendants verbose=true` no longer return the
+  top-level fee fields `fee`, `modifiedfee`, `ancestorfees` and
+  `descendantfees`. These fields were deprecated in v0.17 and moved to the
+  `fees` JSON object in the response. The `-deprecated=fees` configuration
+  option must be passed for these fields to be included in the top-level
+  response. This option will only be available in this major version, after
+  which the fields will be fully removed. Use the `fees` JSON object for these
+  RPCS instead. (#22689)
```
Mostly wording suggestions, but also I think you should warn users about the fact that the fees are denominated slightly differently when they switch to using the `fees` object.
```suggestion
- The top-level fee fields `fee`, `modifiedfee`, `ancestorfees` and `descendantfees`
  returned by RPCs `getmempoolentry`,`getrawmempool(verbose=true)`,
  `getmempoolancestors(verbose=true)` and `getmempooldescendants(verbose=true)`
  are now deprecated and will be removed in the next major version (use
  `-deprecated=fees` if needed in this version). The same values, denominated
  in satoshis, can be accessed through the `fees` object in the result. (#22689)
```

glozow replied:
Doesn't seem to be an issue here, but for future reference, you might want to put it in a release-notes-22689.md to avoid conflicts with other PRs updating release notes.

josibake replied:
> slightly differently when they switch to using the fees object

just double checking, the update should read "denominated in BTC," correct?

glozow replied:
yes, BTC, sorry

glozow reviewed (commented):
utACK, Are we missing tests for `getmempooldescendants` and `getmempoolancestors`?

jonatack commented on src/rpc/blockchain.cpp:
```diff
@@ -490,21 +490,25 @@ static void entryToJSON(const CTxMemPool& pool, UniValue& info, const CTxMemPool
     info.pushKV("ancestorsize", e.GetSizeWithAncestors());
-    info.pushKV("ancestorfees", e.GetModFeesWithAncestors());
+    if (IsDeprecatedRPCEnabled("fees"))
+        info.pushKV("ancestorfees", e.GetModFeesWithAncestors());
```
- brackets per developer-notes.md:
```
  - If an `if` only has a single-statement `then`-clause, it can appear
    on the same line as the `if`, without braces. In every other case,
    braces are required, and the `then` and `else` clauses must appear
    correctly indented on a new line.
```
- `IsDeprecatedRPCEnabled()` loops through a vector to match strings, so maybe invoke it once and store the result

```diff
@@ -493,7 +493,8 @@ static void entryToJSON(const CTxMemPool& pool, UniValue& info, const CTxMemPool
     info.pushKV("vsize", (int)e.GetTxSize());
     info.pushKV("weight", (int)e.GetTxWeight());
     // TODO: top-level fee fields are deprecated. IsDeprecatedRPCEnabled("fees") blocks should be removed in v24
-    if (IsDeprecatedRPCEnabled("fees")) {
+    const bool deprecated_fees_enabled{IsDeprecatedRPCEnabled("fees")};
+    if (deprecated_fees_enabled) {
         info.pushKV("fee", ValueFromAmount(e.GetFee()));
         info.pushKV("modifiedfee", ValueFromAmount(e.GetModifiedFee()));
     }
@@ -501,12 +502,14 @@ static void entryToJSON(const CTxMemPool& pool, UniValue& info, const CTxMemPool
     info.pushKV("height", (int)e.GetHeight());
     info.pushKV("descendantcount", e.GetCountWithDescendants());
     info.pushKV("descendantsize", e.GetSizeWithDescendants());
-    if (IsDeprecatedRPCEnabled("fees"))
+    if (deprecated_fees_enabled) {
         info.pushKV("descendantfees", e.GetModFeesWithDescendants());
+    }
     info.pushKV("ancestorcount", e.GetCountWithAncestors());
     info.pushKV("ancestorsize", e.GetSizeWithAncestors());
-    if (IsDeprecatedRPCEnabled("fees"))
+    if (deprecated_fees_enabled) {
         info.pushKV("ancestorfees", e.GetModFeesWithAncestors());
+    }
```

josibake commented:
pushed changes per @jonatack suggestions and @glozow 's recommendation on the release notes. if the release notes need further refinement, i think we can wait until the draft is moved to the wiki. ive made a note for myself to revisit then.

@glozow , regarding your question on missing tests for `getmempoolancestors` and `getmempooldescendants`: i had started to write tests for these, but since each RPC is calling `entryToJSON` underneath, it didn't seem like there was any added benefit and creating the tests was quite a bit more complicated than the the ones for `getrawmempool` and `getmempoolentry`. as i understand, `rpc_mempool_entry_fee_fields_deprecation.py` will be removed in the in v24 anyways, so i dont see the benefit of writing them, unless others feel strongly that they are needed.

thanks again for the reviews!

jonatack commented on doc/release-notes.md:
```diff
@@ -63,6 +63,12 @@ P2P and network changes
+  `getmempoolancestors(verbose=true)` and `getmempooldescendants(verbose=true)`
+  are now deprecated and will be removed in the next major version (use
+  `-deprecated=fees` if needed in this version). The same values, denominated
+  in BTC, can be accessed through the `fees` object in the result. (#22689)
```
The units don't seem to be changed and remain in BTC/kvB?

Feel free to ignore (the help prints the wrong unit as well, I believe)
```suggestion
  in BTC/kvB, can be accessed through the `fees` object in the result. (#22689)
```

josibake replied:
yeah, i think the help is incorrect here. i can open a general RPC cleanup PR. for starters:

- address where argument order doesn't match help order
- fix fee vs feerate and other small errors

jonatack replied:
So, `fee` and `modifiedfee` are uniformly in BTC/kvB. OTOH, it looks to me like `ancestorfees` and `descendantfees` are in sat/vB in the deprecated fields and in BTC/kvB in the `fees` object (urgh, we've been moving from BTC/kvB to sat/vB elsewhere).

If we mention this in the release notes (and I think it's a good idea as @glozow suggested), it may be best to (a) use the correct units of BTC/kvB and (b) clarify which fields change from sat/vB to BTC/kvB (`ancestorfees` and `descendantfees`) in the release notes.

Finally, here's a commit in https://github.com/jonatack/bitcoin/commits/pr22689-help-fixups that both fixes and improves the docs for these fields, feel free to use or pull it in here since these lines are already being touched (or later, if you prefer) as the helps could use some...help :)

glozow replied:
Note that the units for ancestor and descendant fees are different - they were in satoshis and are now in BTC

jonatack reviewed (commented):
ACK 30563b4bf32b24a25a2f40d2e48e5d64cc79f5c2

josibake commented:
rebased to fix conflicts from #22707 

@MarcoFalke @mjdietzx @jonatack @glozow if everything looks good, a re-ack would be much appreciated! thanks!

josibake commented:
@jonatack thanks for the fix-ups on the field help! i added your suggestions and included you as a co-author in https://github.com/bitcoin/bitcoin/pull/22689/commits/4a6e7b2f0321df876e1d90d06c44325116bf5eec. i also updated the release note docs in https://github.com/bitcoin/bitcoin/pull/22689/commits/61a5b370594b538488dbddc4aadb208c2770d748 to be more clear that `ancestorfees` and `descendantfees` were previously sat/vB and are now `BTC/kvB` in the fees object.

jnewbery commented on src/rpc/blockchain.cpp:
```diff
@@ -490,21 +490,28 @@ static void entryToJSON(const CTxMemPool& pool, UniValue& info, const CTxMemPool
+        info.pushKV("ancestorfees", e.GetModFeesWithAncestors());
+    }
     info.pushKV("wtxid", pool.vTxHashes[e.vTxHashesIdx].first.ToString());
+    info.pushKV("fees", fees);
```
I don't think the last commit needs to be in this PR, since it's not really related to deprecating the legacy fields.

If you do move this down, then it makes sense to also move the preceding lines down so the `fees` code isn't split up:

```c++
    UniValue fees(UniValue::VOBJ);
    fees.pushKV("base", ValueFromAmount(e.GetFee()));
    fees.pushKV("modified", ValueFromAmount(e.GetModifiedFee()));
    fees.pushKV("ancestor", ValueFromAmount(e.GetModFeesWithAncestors()));
    fees.pushKV("descendant", ValueFromAmount(e.GetModFeesWithDescendants()));
```

jnewbery reviewed (commented):
Concept ACK

josibake commented:
thanks for the review @jnewbery ! i added your suggestion in https://github.com/bitcoin/bitcoin/pull/22689/commits/0dd79b51a3de0cc618704077f71aa591cdc8c01e

while i agree moving the fee object is not directly related to fee field deprecation, my (albeit weak) argument for keeping it in this PR is making sure the descriptions are accurate for the affected RPC's will help avoid confusion on the affected user's part as they are looking at the updated descriptions and returned objects and migrating to the new fee fields.

that being said, very easy to pop this commit off and include it in a follow up PR for cleaning up RPC help text discrepancies

jonatack commented on doc/release-notes.md:
```diff
@@ -63,6 +63,14 @@ P2P and network changes
+  `-deprecated=fees` if needed in this version). The same fee rates, all denominated
+  in BTC/kvB, can be accessed through the `fees` object in the result. Note: deprecated
+  fields `ancestorfees` and `descendantfees` are denominated in sat/vB, whereas all
+  fields in the `fees` object are denomainted in BTC/kvB (#22689)
```
```suggestion
  fields in the `fees` object are denominated in BTC/kvB (#22689)
```

perhaps `s/Note:/Warning:`

josibake replied:
oof, good catch. i need to enable spell-check for vim :roll_eyes:

josibake commented:
force pushed to update release notes to fix wording and a spelling error

glozow commented on src/rpc/blockchain.cpp:
```diff
@@ -455,23 +455,23 @@ static RPCHelpMan getdifficulty()
     RPCResult{RPCResult::Type::NUM, "height", "block height when transaction entered pool"},
     RPCResult{RPCResult::Type::NUM, "descendantcount", "number of in-mempool descendant transactions (including this one)"},
     RPCResult{RPCResult::Type::NUM, "descendantsize", "virtual transaction size of in-mempool descendants (including this one)"},
-    RPCResult{RPCResult::Type::STR_AMOUNT, "descendantfees", "modified fees (see above) of in-mempool descendants (including this one) (DEPRECATED)"},
+    RPCResult{RPCResult::Type::STR_AMOUNT, "descendantfees", "transaction fee rate of in-mempool descendants (including this one) in " + CURRENCY_ATOM + "/vB with fee rate deltas used for mining priority (DEPRECATED, returned only if config option -deprecatedrpc=fees is passed)"},
     RPCResult{RPCResult::Type::NUM, "ancestorcount", "number of in-mempool ancestor transactions (including this one)"},
     RPCResult{RPCResult::Type::NUM, "ancestorsize", "virtual transaction size of in-mempool ancestors (including this one)"},
-    RPCResult{RPCResult::Type::STR_AMOUNT, "ancestorfees", "modified fees (see above) of in-mempool ancestors (including this one) (DEPRECATED)"},
+    RPCResult{RPCResult::Type::STR_AMOUNT, "ancestorfees", "transaction fee rate of in-mempool ancestors (including this one) in " + CURRENCY_ATOM + "/vB with fee rate deltas used for mining priority (DEPRECATED, returned only if config option -deprecatedrpc=fees is passed)"},
     RPCResult{RPCResult::Type::STR_HEX, "wtxid", "hash of serialized transaction, including witness data"},
     RPCResult{RPCResult::Type::OBJ, "fees", "",
         {
-            RPCResult{RPCResult::Type::STR_AMOUNT, "base", "transaction fee in " + CURRENCY_UNIT},
-            RPCResult{RPCResult::Type::STR_AMOUNT, "modified", "transaction fee with fee deltas used for mining priority in " + CURRENCY_UNIT},
-            RPCResult{RPCResult::Type::STR_AMOUNT, "ancestor", "modified fees (see above) of in-mempool ancestors (including this one) in " + CURRENCY_UNIT},
-            RPCResult{RPCResult::Type::STR_AMOUNT, "descendant", "modified fees (see above) of in-mempool descendants (including this one) in " + CURRENCY_UNIT},
+            RPCResult{RPCResult::Type::STR_AMOUNT, "base", "transaction fee rate in " + CURRENCY_UNIT + "/kvB"},
+            RPCResult{RPCResult::Type::STR_AMOUNT, "modified", "modified fee rate in " + CURRENCY_UNIT + "/kvB with fee rate deltas used for mining priority"},
+            RPCResult{RPCResult::Type::STR_AMOUNT, "ancestor", "transaction fee rate of in-mempool ancestors (including this one) in " + CURRENCY_UNIT + "/kvB with fee rate deltas used for mining priority"},
+            RPCResult{RPCResult::Type::STR_AMOUNT, "descendant", "transaction fee rate of in-mempool descendants (including this one) in " + CURRENCY_UNIT + "/kvB with fee rate deltas used for mining priority"},
```
`CTxMemPoolEntry.GetModifiedFee()`, `CTxMemPoolEntry.GetFee()`, `CTxMemPoolEntry.GetModFeesWithAncestors()`, `CTxMemPoolEntry.GetModFeesWithDescendants()` return `CAmount`s. See their [declarations](https://github.com/bitcoin/bitcoin/blob/3755dea3c502f32048b110a4868daefe9dbe3cc1/src/txmempool.h#L125-L154) in txmempool.h.

We aren't dividing these fee amounts by sizes, just pushing them directly into the result object in `entryToJSON`.

jonatack replied:
Interesting. In `CTxMemPool::PrioritiseTransaction()`, is `TxMempoolInfo::nFeeDelta` incorrectly described as a fee rate? In BIP22 it's described as a fee. 
```cpp
LogPrintf("PrioritiseTransaction: %s feerate += %s\n", hash.ToString(), FormatMoney(nFeeDelta));
```

josibake replied:
you are correct! my apologies, i should have double-checked the declarations before making the change :grimacing:

glozow commented on doc/release-notes.md:
```diff
@@ -64,6 +64,15 @@ P2P and network changes
+  returned by RPCs `getmempoolentry`,`getrawmempool(verbose=true)`,
+  `getmempoolancestors(verbose=true)` and `getmempooldescendants(verbose=true)`
+  are now deprecated and will be removed in the next major version (use
+  `-deprecated=fees` if needed in this version). The same fee rates, all denominated
+  in BTC/kvB, can be accessed through the `fees` object in the result. Warning: deprecated
```
```suggestion
  `-deprecated=fees` if needed in this version). The same fee, all denominated
  in BTC, can be accessed through the `fees` object in the result. Warning: deprecated
```

glozow reviewed (changes requested):
I'm sorry but I don't think it's correct to relabel these as feerates instead of fees. It's possible there are places where we are confusing the two, but this is not one of those cases.

josibake commented:
as @glozow pointed out, all fee fields returned by `entryToJSON` are absolute amounts (not fee rates). ive update the rpc help and release notes to reflect this (thanks @glozow , sorry for the churn on this).

also kept the wording changes from @jonatack as i think they really improve the readability of the help and i added in the units for each of the fields so it's more obvious that we are changing `ancestorfees` and `descendantfees` from sats to BTC

glozow commented on test/functional/rpc_mempool_entry_fee_fields_deprecation.py:
```diff
@@ -0,0 +1,67 @@
+        self.test_getrawmempool(txid, deprecated_fields)
+        self.test_deprecated_fields_match(txid)
+
+    def test_getmempoolentry(self, txid, deprecated_fields):
+
+        self.log.info("Test getmempoolentry rpc")
+        entry = self.nodes[0].getmempoolentry(txid)
+        deprecated_entry = self.nodes[1].getmempoolentry(txid)
+        assertions_helper(entry, deprecated_entry, deprecated_fields)
+
+    def test_getrawmempool(self, txid, deprecated_fields):
+
+        self.log.info("Test getrawmempool rpc")
+        entry = self.nodes[0].getrawmempool(verbose=True)[txid]
+        deprecated_entry = self.nodes[1].getrawmempool(verbose=True)[txid]
+        assertions_helper(entry, deprecated_entry, deprecated_fields)
+
+    def test_deprecated_fields_match(self, txid):
+
```
nit in e2cada0ce5: these are unnecessary
```suggestion
    def test_getmempoolentry(self, txid, deprecated_fields):
        self.log.info("Test getmempoolentry rpc")
        entry = self.nodes[0].getmempoolentry(txid)
        deprecated_entry = self.nodes[1].getmempoolentry(txid)
        assertions_helper(entry, deprecated_entry, deprecated_fields)

    def test_getrawmempool(self, txid, deprecated_fields):
        self.log.info("Test getrawmempool rpc")
        entry = self.nodes[0].getrawmempool(verbose=True)[txid]
        deprecated_entry = self.nodes[1].getrawmempool(verbose=True)[txid]
        assertions_helper(entry, deprecated_entry, deprecated_fields)

    def test_deprecated_fields_match(self, txid):
```

glozow replied:
newlines still there?

glozow reviewed (commented):
ACK 4792072c50af1d3cefa0899fb914f9306ee3a57c

Doc improvements can be nice, but I personally don't recommend refactoring/rewording things that you intend to remove, as it requires more attention from reviewers but adds little value.

pinging @jonatack, if you have time to take another look?

jonatack commented on src/rpc/blockchain.cpp:
```diff
@@ -455,23 +455,23 @@ static RPCHelpMan getdifficulty()
     RPCResult{RPCResult::Type::NUM, "ancestorcount", "number of in-mempool ancestor transactions (including this one)"},
     RPCResult{RPCResult::Type::NUM, "ancestorsize", "virtual transaction size of in-mempool ancestors (including this one)"},
-    RPCResult{RPCResult::Type::STR_AMOUNT, "ancestorfees", "modified fees (see above) of in-mempool ancestors (including this one) (DEPRECATED)"},
+    RPCResult{RPCResult::Type::STR_AMOUNT, "ancestorfees", "transaction fees of in-mempool ancestors (including this one) with fee deltas used for mining priority, denominated in " + CURRENCY_ATOM + "s (DEPRECATED, returned only if config option -deprecatedrpc=fees is passed)"},
```
only if you retouch, here and line 464
```suggestion
    RPCResult{RPCResult::Type::STR_AMOUNT, "ancestorfees", "transaction fees of in-mempool ancestors (including this one) with fee deltas used for mining priority, denominated in satoshis (DEPRECATED, returned only if config option -deprecatedrpc=fees is passed)"},
```

jonatack reviewed (commented):
ACK 4792072c50af1d3cefa0899fb914f9306ee3a57c

modulo some release note fixups and maybe `s/sats/satoshis/`.

I'm glad to see the helps for the four RPCs updated consistently/coherently for both the non-deprecated and deprecated fields, as otherwise it could be confusing for users of 0.23, both while it is the newest release and also afterward for as long as users can run it, which could be a very long time.

Sorry for the time to re-review, will be quicker if you retouch as the changes should be very easy to re-review.

jnewbery reviewed (commented):
The set of changes look fine, but there a couple of issues with commits:

- The mempool_packages.py fails after the first commit since you remove the `COIN` import (and then add it back in the second commit). I think you just need to squash the changes to mempool_packages.py in the second commit into the first commit.
- Can you fixup the commit logs with standard punctuation/capitalization, backticks to indicate code, full filenames, etc.

josibake commented:
> The set of changes look fine, but there a couple of issues with commits:
> 
>     * The mempool_packages.py fails after the first commit since you remove the `COIN` import (and then add it back in the second commit). I think you just need to squash the changes to mempool_packages.py in the second commit into the first commit.
> 
>     * Can you fixup the commit logs with standard punctuation/capitalization, backticks to indicate code, full filenames, etc.

thanks for taking a look at this again, @jnewbery ! looks like i added back the `COIN` import as part of the rebase but added it to the wrong commit. fixed now. also cleaned up the commit messages and added the newline per your suggestion.

`git range-diff 285cb3d...7330beb`

jnewbery commented on test/functional/rpc_mempool_entry_fee_fields_deprecation.py:
```diff
@@ -0,0 +1,67 @@
+        self.wallet = MiniWallet(self.nodes[0])
+        self.wallet.rescan_utxos()
+
+        # we create the tx on the first node and wait until it syncs to node_deprecated
+        # thus, any differences must be coming from getmempoolentry or getrawmempool
+        txid = self.wallet.send_self_transfer(from_node=self.nodes[0])["txid"]
+        self.sync_mempools(nodes=self.nodes)
```
This sync can take several seconds since bitcoind will add some random delay on transaction relay to improve privacy. Since testing tx relay isn't the point of this test, you could make the test run reliably faster by submitting the transaction directly to node 1 with the `sendrawtransaction` RPC.

josibake replied:
thanks for the tip! i did notice the new test was taking longer than i expected. fixed!

josibake replied:
change since your last ACK: `git range-diff  7330beb...2f9515f`

jnewbery reviewed (commented):
Code review ACK 7330beb391

One suggestion inline to speed up the new function test. Thanks for the quick turnaround @josibake!

jnewbery commented:
reACK 2f9515f37addabde84c79926d7a24b2897a21dd1

One minor comment above: https://github.com/bitcoin/bitcoin/pull/22689#discussion_r740930133. No need to resolve that unless you need to push again for some other reason.

glozow reviewed (commented):
utACK 2f9515f37addabde84c79926d7a24b2897a21dd1

MarcoFalke commented on src/rpc/blockchain.cpp:
```diff
@@ -462,23 +462,23 @@ static RPCHelpMan getdifficulty()
-    RPCResult{RPCResult::Type::STR_AMOUNT, "fee", "transaction fee in " + CURRENCY_UNIT + " (DEPRECATED)"},
-    RPCResult{RPCResult::Type::STR_AMOUNT, "modifiedfee", "transaction fee with fee deltas used for mining priority (DEPRECATED)"},
+    RPCResult{RPCResult::Type::STR_AMOUNT, "fee", "transaction fee, denominated in " + CURRENCY_UNIT + " (DEPRECATED, returned only if config option -deprecatedrpc=fees is passed)"},
+    RPCResult{RPCResult::Type::STR_AMOUNT, "modifiedfee", "transaction fee with fee deltas used for mining priority, denominated in " + CURRENCY_UNIT + " (DEPRECATED, returned only if config option -deprecatedrpc=fees is passed)"},
```
missing optional. see https://github.com/bitcoin/bitcoin/pull/23694
